(repost) - Obtain version on class init, allow compatibility with 2.x and 3.x APIs.

jasonbarbee opened the pull request:
Per comments of PR #151, @falkowich you asked me to recommit.  I re-synced with master, and tried to re-commit.  Pardon the "new line" I couldn't seem to match that, I didn't want to change any code there, it seems some kind of editor issue.

JonasKs commented:
This should be optional. We use ISE through APIs and have no desire to do an extra API call for every request that hit our API (new class is instanced for each request)

JonasKs commented on ise.py:
```diff
@@ -169,6 +171,22 @@ def _request(self, url, method="get", data=None):
+            short_version = float(full_version[0:3])
+            # print("ISE Initializing - Version Check " + full_version)
+            return short_version
+        except:
```
Really not a fan of this broad exception. If there is a specific exception expected due to ISE version, that specifically should be caught.

jasonbarbee replied:
Keeping it simple - throwing an exception would break existing compatibility, as this PR maintains. As proposed, the library init code in this PR just continues as it does today for all users of the library. If bad data or credentials were entered, today the library only fails deeper in the code - on actual querying of data. This causes a library user to put more guardrails further down in the code and catch 403s when they were expecting API results. This change provides a simple way to check credentials, and if the Monitoring API was reachable, and fill in the version if available (it's a sanity check before calling an operation against the ERS REST API).  

Also note that ISE ERS Rest API in a multi node cluster is read-only if the Admin node is Standby. It's best to query ISE Monitoring API, to find which Admin node has the active role. If the library makes PUT request to a standby node, it will fail. This PR lays groundwork to get the active admin node from the MnT API, and makes the library more friendly to use by saving users from tripping and then trying to code around ISE API pitfalls like this. The only cost is a simple credential validation and query against the Monitoring API, before actually using the ERS REST API. I'll code that discovery in the next PR.

As an added bonus to confirming API credentials early, (later) obtaining the Active node, you get to discover the API version in the transaction. Knowing the version at class init would provide a way to block invalid API requests from being sent (sending a 3.x API request to a 2.x server, and vice versa). Or if a parameter changes between versions like we saw in #148, it provides a way to adapt the library to the API versions as they are released. I think the library has grown to a point that it needs a way to handle API version changes.

I think the concern you have is a small initialization time cost, but there are many benefits list above, when users will be counting on the accuracy and resiliency of this library rather than shaving off a few ms to turn off safety checks. If it's changed to be optional, I think it would be better on by default, and could be disabled if a user wants all the safeguards off.

JonasKs replied:
> I think the library has grown to a point that it needs a way to handle API version changes.

I disagree on this one. I’ve never seen an SDK that fetches things behind the scenes to determine what API version you want to reach.  A setting that we can pin, is what I expect and what I would want. For those in a situation where ISE gets patched without them knowing or testing against the patch, that sounds like a whole different issue. 

I understand the intent here, I just don’t think it should do all this without opting in. Adding “a small initialization time cost” to an SDK makes it less attractive for those who builds APIs using it. I understand scripts and scheduled tasks could benefit from this, but this is a downgrade in behavior for me.

If this is wanted in the library, I suggest it’s added as an optional setting. If it’s on by default I don’t really care, though I would argue I’d major bump it to give everyone a heads up.   
I don’t want my code (against a security system, especially) to be able to change dynamically without me knowing about it(or specifically asking for it). I’d rather have it fail. 


With all this said, you should the very least catch the error and log it as an exception.

JonasKs replied:
All in all, I think both of us will be happy with a setting for this. Whether it’s on by default or not don’t really matter to me personally, that’s something Falk probably should think about. 😊

joshand replied:
+1 that this should be optional. I don't have a problem with it being on by default, but please give me a way to opt out of the automatic version check.

jasonbarbee commented:
After thinking on this more... I split up the code into 2 optional parameters, and went ahead and added the failover support to this PR (hope that's not too much, but it seemed be related so we could all see how it looks)

Version Discovery - on by default, toggle by setting the version on init:
Default version variable, set to '', if it's blank we discover it, if it is set on initialization - we use that, and skip version discovery. The user can choose to lock in a version, or if it is blank - it's automatic discovery of the version. 

HA Discovery off by default, toggle on with a boolean variable:
parameter - enable_failover (true/false) - takes care of the HA discovery features. Default to false - because not everyone needs this. 

Possible improvement to do is that the library only takes 1 IP today. The failover discovery will work as long as that IP has either Active or Standby role. If it's standby, it will find the other primary node, and assign it to the class on init, so all future calls will use the discovered IP. I could just add a new parameter - secondary_ise_node, or ise_node2, and run queries there if the primary is offline.

Thoughts?

JonasKs commented on ise.py:
```diff
@@ -57,9 +65,78 @@ def __init__(
+                        print("Error: Connection Error to Secondary ISE Node.")
+                        raise
+            except requests.exceptions.RequestException as e:
+                print("Error: Connection Error to Secondary ISE Node.")
```
All prints should be converted into logging statements. 

```py
import logging

log = logging.getLogger('ise')
log.debug(…)
log.warning(…)

```

jasonbarbee replied:
Added logging to replace prints.

falkowich commented:
I will check this out as soon as #164 is done

And I am sorry for taking so much time with this..

--
Regards Falk

falkowich commented:
If there is a dedicated ERS user this will not work i guess?

How is the best way to fix that?  
Another user_pass dict so that we can use a read_only user that can scrape the frontend?  

Beq cisco wanted a namechange I had to do #164.  
So your merge is in bransh https://github.com/falkowich/pyise-ers/pull/168

--
Regards Falk

jasonbarbee commented:
Yeah, they need both MnT and ERS group rights in current code, which may be a little unexpected. I have an idea.

I just found an ISE ERS API call that will let us query if the ERS node is primary or not, and even see the secondary. I can move this feature down to the ERS only permission. 
https://developer.cisco.com/docs/identity-services-engine/3.0/#!node-details/get-by-id

I can build changes to use ERS only, and fix the conflict - any other changes requested?

falkowich commented:
Any more information on this, I'm trying to clean up some PR's before 0.3.
So if someone have more information on this I'm all ears :)

falkowich commented:
Closing this.
If it's still up for merge, please renew the PR.

Really sorry for the trouble.
